### ISSUE TITLE
[Bug]: Jovian Multi-Row CUDA-graph NaN breaking C>=2

### ISSUE THREAD

**Issue by swein** [local-inference-lab/vllm#621]
### Your current environment

- Image: `voipmonitor/vllm:jovian-judgement-community-20260902-r18`
- vLLM source: in-tree at `/opt/glm53-flash/vllm` (PYTHONPATH-over-venv), git `a4b04eea0` + ~90 uncommitted files (Luke's NVFP4/Flash-Next lineage, b12x `ab7f240bdc`);
engine version string is stale, don't trust it
- Model: `Qwen3.8-Flash-Next-NVFP4-4p89` (hybrid GDN linear-attn + full-attn, NVFP4/MXFP4, MTP draft layer)
- Hardware: RTX PRO 6000 Blackwell 96GB, TP=1
- Flags: `--max-num-seqs 16 --block-size 16 --mamba-cache-mode align --gdn-decode-kernel b12x
  --kv-cache-dtype fp8_e4m3 --quantization modelopt_mixed`, `--speculative-config '{"method":"mtp","num_speculative_tokens":3}'`

### 🐛 Describe the bug

Root cause

 Padded FULL-graph replay. The dispatch layer pads a uniform decode batch up to the next captured descriptor (rows and tokens). On r18, at least one kernel inside the
 captured region mishandles the phantom padded row's stale bindings and corrupts state — target side: instant 100%-NaN logits for all rows; draft side: poisoned
 recurrent-state pages, acceptance pinned at 1.00, later requests inherit NaN via reused pages.

 Why C=1 was healthy and C≥2 died:
 - Target decode graphs rounded capture sizes to multiples of decode_query_len=4 → set was (4,1),(8,2),(16,4),(24,6)…(64,16). 1-row batches hit (4,1) exactly — zero
   padding. 3-row batches padded 3→4 rows (12→16 tokens).
 - The MTP draft-prefill manager additionally filters capture sizes to powers of two (_sparse_full_capture_request_sizes → {1,2,4,8,16}), so every non-power-of-two
   multi-row draft prefill padded (proved by B1: exact-size target clean but acceptance 1.00 → drafts all rejected → canary NaN ~60 s later via page poisoning).

 Bisect boots (all prod profile, MTP3, disk offload, tripwires armed):

``` ┌──────┬─────────────────────────────────────────────────┬──────────────┬────────────────────────────────────────────┐
 │ Boot │ Change                                          │ cold 3       │ Verdict                                    │
 ├──────┼─────────────────────────────────────────────────┼──────────────┼────────────────────────────────────────────┤
 │ B1   │ dense target sizes incl. (12,3),(20,5)          │ clean, 0 NaN │ padding is the trigger; draft still padded │
 ├──────┼─────────────────────────────────────────────────┼──────────────┼────────────────────────────────────────────┤
 │ B2   │ stock sizes + VLLM_CAUSAL_CONV1D_UPDATE_HOIST=0 │ crash        │ image-only hoisted conv path exonerated    │
 ├──────┼─────────────────────────────────────────────────┼──────────────┼────────────────────────────────────────────┤
 │ B4   │ dense sizes everywhere (target + draft)         │ clean        │ fixed                                      │
 └──────┴─────────────────────────────────────────────────┴──────────────┴────────────────────────────────────────────┘
```

 The kernel-level padded-row defect itself remains open upstream; statically verified padded-safe: causal_conv1d_update (qs==qe + NULL guards), triton
 fused_sigmoid_gating_delta_rule_update (T==0, state_idx<=0), GDN builder buffer tails, FA scheduler staging. Confirmed related latent bug for Luke:
 qwen_gdn_linear_attn.py _run_b12x_gdn_decode_post_conv ~line 1996 — self._b12x_num_seqs.fill_(num_requests) bakes the capture-time python row count into the graph (all
 sibling bindings are GPU copies). Harmless once graphs are exact-size; wrong under padding. GPU-safe derivation: (state_indices[:cap,0] > 0).sum().

 Shipped fix (verified live now)

 /mnt/speedteam/models/vllm-server-38-next-lukemix.sh (prod profile):
 1. --compilation-config '{"cudagraph_capture_sizes":[1..24,28,32,…,64]}' — every 1–16-row MTP3 step replays an exact-size graph (NO_CUDAGRAPH=1 still overrides; graph
    mem 0.79+0.42 GiB vs 0.53+0.32).
 2. Sideload bind: /mnt/speedteam/models/patch-r18/.../speculator.py → /opt/glm53-flash/vllm/vllm/... (dense draft-prefill sizes; remove when the image ships it).
 3. NaN tripwire ON by default (NAN_TRIPWIRE=0 disables).

 Verification (exact ask: original lukemix + KV cpu offload + concurrency + MTP3)

```
 diff --git a/vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py b/vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py
index 64ba53c..3d85b27 100644
--- a/vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py
+++ b/vllm/v1/worker/gpu/spec_decode/autoregressive/speculator.py
@@ -27,14 +27,14 @@ logger = init_logger(__name__)


 def _sparse_full_capture_request_sizes(max_num_reqs: int) -> frozenset[int]:
-    """Return power-of-two request capacities, including the configured maximum."""
-    request_sizes = []
-    request_size = 1
-    while request_size < max_num_reqs:
-        request_sizes.append(request_size)
-        request_size *= 2
-    request_sizes.append(max_num_reqs)
-    return frozenset(request_sizes)
+    """Return the request capacities draft prefill FULL graphs are captured at.
+
+    Padded-row replay is broken on this build: a batch whose request count
+    pads up to a captured size corrupts recurrent-state pages (r18 multi-row
+    NaN). Capture at EVERY request size up to max_num_reqs so draft prefill
+    steps always replay an exact-size graph.
+    """
+    return frozenset(range(1, max_num_reqs + 1))
```

No existing issue in local-inference-lab/vllm tracks this bug. Nothing covers padded FULL-graph replay NaN on multi-row batches (100% NaN logits / poisoned
 recurrent-state pages / acceptance pinned at 1.00, fixed by dense capture sizes), and nothing tracks the _b12x_num_seqs capture-time bake-in either. A new issue would
 not be a duplicate.

 Coverage of the check: read every issue+PR #416–#616 (Aug 17 → today) individually, plus repo-wide searches (NaN in:title → 0 results across the whole repo; full-text
 corrupted NaN graph draft, padded graph replay, MTP draft NaN). Items older than #416 predate the r18 image and surfaced nothing on padded-replay corruption beyond
 what's listed below.

 Closest existing items, none a match:

``` ┌─────────────────────────────────────────┬──────────────────────────────────────────────────┬──────────────────────────────────────────────────────────────────────────┐
 │ Item                                    │ What it is                                       │ Why it's not this bug                                                    │
 ├─────────────────────────────────────────┼──────────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────┤
 │ #616                                    │ GLM-5.3-Flash TP4 DGX Spark: decode corrupted by │ Same family (multi-row decode corruption, GLM-5.3-Flash, spec decode),   │
 │ (https://github.com/local-inference-lab │ concurrent chunked prefill; single-stream clean, │ but wrong-but-well-formed answers, not NaN; their PIECEWISE + no-spec    │
 │ /vllm/issues/616) (open, today)         │ multi-stream corrupted; rate scales with         │ boot still failed, which rules out padded graph replay as its cause;     │
 │                                         │ --max-num-batched-tokens                         │ different HW (GB10), variant (NVFP4-Spark), draft (DFlash2 K7 vs MTP3)   │
 ├─────────────────────────────────────────┼──────────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────┤
 │ #590                                    │ r17 qualified B12X/vLLM merge checklist for      │ Not a bug report, but the landing zone context: your speculator.py       │
 │ (https://github.com/local-inference-lab │ GLM-5.3-Flash TP4                                │ dense-sizes fix and the _b12x_num_seqs bug sit exactly on this stack     │
 │ /vllm/issues/590) (open)                │                                                  │                                                                          │
 ├─────────────────────────────────────────┼──────────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────┤
 │ #502                                    │ perf(spec-decode): bound autoregressive draft    │ This is almost certainly the PR that introduced the power-of-two         │
 │ (https://github.com/local-inference-lab │ prefill graph memory                             │ _sparse_full_capture_request_sizes filter your patch reverts — the fix   │
 │ /vllm/pull/502) (open PR)               │                                                  │ gives back ~0.36 GiB of the memory it saved (0.79+0.42 vs 0.53+0.32      │
 │                                         │                                                  │ GiB); upstream needs to reconcile the two                                │
 ├─────────────────────────────────────────┼──────────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────┤
 │ #600                                    │ Qwen3.8 PLE state corruption under piecewise     │ Same class of graph-state corruption in the GDN/linear-attention family; │
 │ (https://github.com/local-inference-lab │ CUDA graphs                                      │ different model and graph mode                                           │
 │ /vllm/pull/600) (closed PR)             │                                                  │                                                                          │
 ├─────────────────────────────────────────┼──────────────────────────────────────────────────┼──────────────────────────────────────────────────────────────────────────┤
 │ #478                                    │ Multi-stream cross-script corruption /           │ Different model, spec method, no NaN                                     │
 │ (https://github.com/local-inference-lab │ repetition / acceptance collapse on              │                                                                          │
 │ /vllm/issues/478), #475, rtx6kpro#72    │ DeepSeek-V4-Flash DSpark                         │                                                                          │
 └─────────────────────────────────────────┴──────────────────────────────────────────────────┴──────────────────────────────────────────────────────────────────────────┘
```
### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.
